File: cpp/ReverseArray.cpp
#include <iostream>
#include <array>
using namespace std;

template <typename T, size_t N>
void printArray(array<T, N> inputArray) {
	for (size_t i = 0; i < inputArray.size(); i++) {
		cout << inputArray[i] << " ";
	}
	cout << endl;
}

template <typename T, size_t N>
void reverseArray(array<T, N>& inputArray) {

	int tempValue; // create variable to store temporary value
	size_t rightIndex{ inputArray.size() - 1 }; // right index

	// because this algorithm works from both ends inward toward the center
	// the maximum number of iterations is half of the array length
	for (size_t leftIndex{ 0 }; leftIndex < inputArray.size() / 2; leftIndex++) {
		// store the value of the left index
		tempValue = inputArray[leftIndex];
		// switch value of left index to value of right index
		inputArray[leftIndex] = inputArray[rightIndex];
		// switch value of right index to stored value of left index
		inputArray[rightIndex] = tempValue;
		// decrement the right index
		rightIndex--;

		// for loop iteration increments leftIndex value
	}
}

int main()
{
	// create and initialize the array
	array<int, 25> intArray;
	for (size_t i{ 0 }; i < intArray.size(); i++) {
		intArray[i] = i + 1;
	}

	// display the length of the array
	cout << "\nArray length: " << intArray.size() << endl;
	// print original array order
	printArray(intArray);
	// reverse the array order
	reverseArray(intArray);
	// print reversed array
	cout << "\nReversed array: " << endl;
	printArray(intArray);

}

File: java/ReverseArray.java
public class ReverseArray {

    public static void reverseArray(int inputArray[])
    {
        // create variables for algorithm
        int leftValue;
        int leftIndex = 0;

        // because the algorithm works form both ends of the array inward,
        // the maximum number of iterations is half of the array length
        // Note: this could also be achieved with a for loop
        while (leftIndex < inputArray.length / 2) {
            // store left value for later swap
            leftValue = inputArray[leftIndex];
            // assign the leftIndex the value of the corresponding right index
            inputArray[leftIndex] = inputArray[inputArray.length - leftIndex - 1];
            // assign the right index the stored value of the corresponding
            // right index (the one stored in the leftValue variable)
            inputArray[inputArray.length - leftIndex - 1] = leftValue;
            // increment the counter
            leftIndex++;
        }
    }

    public static void displayArray(int inputArray[]) {
        for (int value : inputArray) {
            System.out.print(value + " ");
        }
    }

    public static void main(String[] args) {

        int exampleArray[] = {1,2,3,4,5,6,7,8,9,10};

        // display original order of the array
        System.out.println("Original order:");
        displayArray(exampleArray);

        // reverse array using method
        reverseArray(exampleArray);

        // display the reversed array
        System.out.println("\nReversed array order:");
        displayArray(exampleArray);
    }
}
